### PR TITLE
Add very early support for Mac Catalyst.

### PR DESCRIPTION
**Pull request by rolfbjarne**
* Install the Mac Catalyst versions of the mono libraries and BCL.
   * The BCL is the same as the one for Xamarin.iOS, which means it has to be post-processed a bit to work with a Xamarin.MacCatalyst.dll
* Build our runtime for Mac Catalyst.
* Build a Xamarin.MacCatalyst.dll with the Mac Catalyst API (it compiles, but I haven't looked at the API surface at all). This PR assumes we're going to have a new TargetFrameworkIdentifier for Mac Catalyst, but a final decision has not been made (see https://github.com/dotnet/runtime/issues/44882), so this may change.
* Build a Xamarin.iOS.dll that contains type forwarders to Mac Catalyst for all the types that exist in both Mac Catalyst and Xamarin.iOS.
* Add support to xharness for running introspection on Mac Catalyst (there are a lot of failures because the API surface is wrong)
* Add support to our msbuild tasks and mtouch for building Mac Catalyst apps. This basically comes down to adding a new case in numerous places to either do things the iOS way or the macOS way, depending on each case.
* Add a \_\_MACCATALYST\_\_ define (which is in addition to the \_\_IOS\_\_ define).

This PR is best reviewed commit-by-commit.

### CODE REVIEW

File: src/Makefile [hunk 1]
```diff
@@ -1187,6 +1188,276 @@ TVOS_TARGETS += \
...
+MACCATALYST_GENERATED_DEFINES= -d:__MACCATALYST__ -d:XAMCORE_3_0 -d:IOS
+
+MACCATALYST_CORE_WARNINGS_TO_FIX=-nowarn:219,618,114,414,1635,3021,$(IOS_WARNINGS_THAT_YOU_SHOULD_FIX)
+MACCATALYST_CORE_DEFINES=$(MACCATALYST_DEFINES) -d:COREBUILD
+MACCATALYST_GENERATOR_DEFINES = $(MACCATALYST_GENERATED_DEFINES)
+MACCATALYST_WARNINGS_TO_FIX=-nowarn:219,618,114,414,1635,3021,$(IOS_WARNINGS_THAT_YOU_SHOULD_FIX)
+
+MACCATALYST_EXTRA_CORE_SOURCES = \
+	$(MACCATALYST_BUILD_DIR)/Constants.cs    \
+    $(MACCATALYST_BUILD_DIR)/AssemblyInfo.cs \
```
Comment (spouliot) [inline src/Makefile:1208]: indent

File: src/Metal/MTLEnums.cs [hunk 2]
```diff
@@ -1089,16 +1089,16 @@ public enum MTLMultisampleStencilResolveFilter : ulong
...
+	[Native]
 	public enum MTLSparseTextureRegionAlignmentMode : ulong {
 		Outward = 0x0,
 		Inward = 0x1,
 	}
 
-	[Unavailable (PlatformName.MacCatalyst)]
 	[Flags, Mac (11,0), NoTV, iOS (13,0)]
+	[MacCatalyst (14, 0)]
 	[Native, Advice ("This API is not available when using UIKit on macOS.")]
```
Comment (spouliot) [inline src/Metal/MTLEnums.cs:1102]: Advice should be removed

File: src/videosubscriberaccount.cs [hunk 3]
```diff
@@ -126,6 +127,7 @@ interface VSAccountManager {
 		IVSAccountManagerDelegate Delegate { get; set; }
 
 		[Unavailable (PlatformName.UIKitForMac)][Advice ("This API is not available when using UIKit on macOS.")]
```
Comment (spouliot) [inline src/videosubscriberaccount.cs:129]: ^ should be removed

**Review by spouliot — COMMENTED** [review]
got a checklist of the remaining todo ? that way I won't comment on too obvious things :)

**Comment by praeclarum** [comment]
Beautiful!!!

**Comment by rolfbjarne** [comment]
> got a checklist of the remaining todo ? that way I won't comment on too obvious things :)

All I did API/binding wise was to make it build, so there's pretty much everything left to do there.

Introspection runs, but there's a lot of failures that must be fixed.

All the other test suites are pending (xtro is on the top of the list).

The type-forwarding Xamarin.iOS.dll needs work with regards to types/API that's not in Mac Catalyst.

**Review by mandel-macaque — APPROVED** [review]

**Review by tj-devel709 — APPROVED** [review]

**Review by emaf — APPROVED** [review]

**Comment by rolfbjarne** [comment]
Test failures:

* [NUnit] Mono SystemNumericsTests/watchOS 32-bits - simulator/Debug: https://github.com/xamarin/xamarin-macios/issues/10211
* [NUnit] Mono MonoRuntimeTests/watchOS 32-bits - simulator/Debug: https://github.com/xamarin/xamarin-macios/issues/10211
* [xUnit] Mono SystemNumericsXunit/watchOS 32-bits - simulator/Debug: https://github.com/xamarin/xamarin-macios/issues/10211
* [xUnit] Mono MicrosoftCSharpXunit/watchOS 32-bits - simulator/Debug: https://github.com/xamarin/xamarin-macios/issues/10211
* [xUnit] Mono SystemCoreXunit Part 1/watchOS 32-bits - simulator/Debug: https://github.com/xamarin/xamarin-macios/issues/10211
* MTouch tests/NUnit: fixed in #10198

**Review by spouliot — APPROVED** [review]
I'd really like a checklist (maybe on the related issue for the feature?) because in the past we often forgot to complete small parts (when something big interrupted large work) only to found out much later.

_yeah, like .framework not fully working on XM being found when doing .xcframework work_ :)

**Comment by rolfbjarne** [comment]
> I'd really like a checklist (maybe on the related issue for the feature?) because in the past we often forgot to complete small parts (when something big interrupted large work) only to found out much later.

I created issues with a `Mac Catalyst` milestone: https://github.com/xamarin/xamarin-macios/milestone/44

**Comment by pooran** [comment]
Any update?

**Comment by rolfbjarne** [comment]
@pooran update on what? This pull request has been merged.

**Comment by pooran** [comment]
I was hoping to know detailed steps to enable our current Xamarin.Forms iOS project for Mac, with this commit being merged

**Comment by rolfbjarne** [comment]
> I was hoping to know detailed steps to enable our current Xamarin.Forms iOS project for Mac, with this commit being merged

There's some documentation here: https://github.com/xamarin/xamarin-macios/wiki/Mac-Catalyst-(Early-Preview)

**Pull request merged**